### ISSUE TITLE
How to setup prod and dev hosting?

### ISSUE THREAD

**Issue by rygo6** [aws-amplify/amplify-cli#128]
What is the best practice for using amplify-cli to deploy to dev and prod environments?

**Comment by swaminator** [comment]
@rygo6 I'm a PM on the team and would love to chat with you to share our plans on how to handle multiple environments. Let me know your email and I'll set something up.

**Comment by alukach** [comment]
@swaminator I'd like to have such a conversation as well (have also read over https://github.com/aws-amplify/amplify-js/issues/875 which seems similar).

You're welcome to send me an email at `anthonylukach@gmail.com`

**Comment by samueljseay** [comment]
@swaminator I'm also interested in this conversation for our team who are looking at doing static hosting on Amplify. my email is `sam@atomic.io`

**Comment by SagarMhatre** [comment]
@swaminator I'd like to have such a conversation as well
You're welcome to send me an email at sagarmmhatre@yahoo.co.in

**Comment by swaminator** [comment]
@SagarMhatre @samueljseay @alukach I've sent you all emails. Please do have a look.

**Comment by gcid12** [comment]
@swaminator , can you also include me in the mail  gcid@artcodedata.com    thanks

**Comment by Myrberg** [comment]
I'll appriciate being included as well @swaminator , with my email dmyrberg@gmail.com

**Comment by sp-tenovos** [comment]
@swaminator My team is very interested in how to handle this as well. shane@tenovos.com

**Comment by hisham** [comment]
Me too: halshura@gmail.com

**Comment by rygo6** [comment]
Is there any word on an official workflow being put out for this?

**Comment by wshaddix** [comment]
I'm interested as well. wshaddix_at_web.com

**Comment by ChrisWun** [comment]
@swaminator Me and my Team is also interested.
c.wunder@hochzehn.com

**Comment by brianhilst** [comment]
@swaminator Is there a workaround for this either official or not? We're trying to decide whether to go with Amplify or not on a new project.

**Comment by rygo6** [comment]
TBH, I would not use amplify on a major project yet. Lots of potential. But needs some more development. If it's a smaller project where you can deal with something more expirimental, and not everything needs to be perfect. Probably would be good, then you can give them some feedback. Tons of great ideas and great direction, but needs more development.

**Comment by CodySwannGT** [comment]
sign me up - cody@gunnertech.com

**Comment by swaminator** [comment]
Awesome to see the interest! I'm going to get in touch with all of you this week.

**Comment by swaminator** [comment]
> TBH, I would not use amplify on a major project yet. Lots of potential. But needs some more development. If it's a smaller project where you can deal with something more expirimental, and not everything needs to be perfect. Probably would be good, then you can give them some feedback. Tons of great ideas and great direction, but needs more development.

@rygo6 would you be open to chatting? Please share your email.

**Comment by alphara** [comment]
I'm curious too: artarakcheev@gmail.com

**Comment by rygo6** [comment]
> > TBH, I would not use amplify on a major project yet. Lots of potential. But needs some more development. If it's a smaller project where you can deal with something more expirimental, and not everything needs to be perfect. Probably would be good, then you can give them some feedback. Tons of great ideas and great direction, but needs more development.
> 
> @rygo6 would you be open to chatting? Please share your email.

sure, email is rygo6@me.com

**Comment by austinamorusocfc** [comment]
Hi I'm super interested in how to accomplish this too. 
aamoruso@cfchildren.org

**Comment by southerneer** [comment]
Me too: eric@southerneer.com

**Comment by kaustavghosh06** [comment]
@rygo6 @alukach @SagarMhatre @gcid12 @Myrberg @chrisfowler @sp-tenovos @cgreening @hisham @wshaddix @ChrisWun @brianhilst @JustFly1984 @CodySwannGT @alphara @austinamorusocfc @southerneer 
We're currently brainstorming and working towards providing a CI/CD workflow as a part of the Amplify CLI, and in doing so we want to work backwards from the customer needs and we've opened up an issue (https://github.com/aws-amplify/amplify-cli/issues/286) as a part of the CLI repo to track this story. 
Would be great, if you guys could take a look at it and respond back with some answers to the questions mentioned in the issue (https://github.com/aws-amplify/amplify-cli/issues/286) which would give us a better idea about your needs. Thanks!

**Comment by kaustavghosh06** [comment]
@ahmetcetin We opened up an RFC as a part of this repo to gather information from customers.
Could you please respond to the same at https://github.com/aws-amplify/amplify-cli/issues/286

**Comment by kaustavghosh06** [comment]
Based on all of your feedback, we’ve released a beta version of the CLI (npm install -g @aws-amplify/cli@multienv)  to support multiple environments, team workflows and integrate with your CI/CD flows.
Please check out this documentation for more details - https://aws-amplify.github.io/docs/cli/multienv?sdk=js
Also, as a part of our feedback mechanism, we’ve opened up another [GitHub issue](https://github.com/aws-amplify/amplify-cli/issues/508),  to track feedback from everyone, iterate on this beta version and then merge it to master.  We'll close this issue and track feedback in #508. Looking forward to your feedback on this release.

**Comment by alphara** [comment]
That's awesome news about the release with multiple environments!

**Comment by alexdorand** [comment]
I am very much interested as well alexdorand@gmail.com, could you please include me as well?

**Comment by kaustavghosh06** [comment]
@alexdorand We've launched support for this in beta and made it publicly available for feedback.
Please checkout the the following announcement:

Based on all of your feedback, we’ve released a beta version of the CLI (npm install -g @aws-amplify/cli@multienv) to support multiple environments, team workflows and integrate with your CI/CD flows.
Please check out this documentation for more details - https://aws-amplify.github.io/docs/cli/multienv?sdk=js
Also, as a part of our feedback mechanism, we’ve opened up another GitHub issue, to track feedback from everyone, iterate on this beta version and then merge it to master. We'll close this issue and track feedback in #508. Looking forward to your feedback on this release.

**Comment by swaminator** [comment]
The Amplify CLI now supports team workflows. If you're looking to use the Amplify CLI in conjunction with the Amplify Console, you can view the recommended workflow here: https://docs.aws.amazon.com/amplify/latest/userguide/multi-environments.html

**Comment by puneetrastogi90** [comment]
Hi, I would also like to set up the environments for my project. @swaminator please help me. my email id is puneetrastogi90@gmail.com

**Comment by zacharycarter** [comment]
I don't think this issue should be closed - it still doesn't seem possible to publish multiple hostings, multienv or not... Unless the answer is - it isn't possible with the CLI and you need to use the console.

**Comment by edpike109** [comment]
You have to install the lastest version of the amplify/cli, latest version
of vanilla amplify does not have latest cli last i checked. Latest cli has
multienv.

**Comment by orland1to** [comment]
Hi, I would also like to set up the environments for my project. @swaminator please help me. my emai is s4lvadorcamarill0@gmail.com

**Comment by dreadjr** [comment]
@swaminator could you share as well? dreadjr@gmail.com

**Comment by swaminator** [comment]
Just to update @dreadjr @orland1to @puneetrastogi90. The Amplify CLI and Amplify Console now support multiple envs.  If you're looking to use the Amplify CLI in conjunction with the Amplify Console, you can view the recommended workflow here: https://docs.aws.amazon.com/amplify/latest/userguide/multi-environments.html

**Comment by AshkanHovold** [comment]
Is it possible to have one environment (dev) that only uses s3 and http and another enviornment (prod) that uses cloudfront and https? I have created different environments but if I do "amplify hosting configure" in one environment the changes effect the other environment has well

**Comment by rmcvicar** [comment]
I also see this issue.  We (currently) don't allow external access to our source code so hooking into the AWS CI/CD workflow isn't really a fit for us at the moment.

> Is it possible to have one environment (dev) that only uses s3 and http and another enviornment (prod) that uses cloudfront and https? I have created different environments but if I do "amplify hosting configure" in one environment the changes effect the other environment as well

**Comment by swaminator** [comment]
> Is it possible to have one environment (dev) that only uses s3 and http and another enviornment (prod) that uses cloudfront and https? I have created different environments but if I do "amplify hosting configure" in one environment the changes effect the other environment as well

Starting from your `dev env:

1. `amplify env add` create`prod` which is a clone of `dev` at this point.
2. `amplify hosting remove`
3. `amplify hosting add` and add CloudFront this time round.

Let me know if that works.

**Comment by swaminator** [comment]
> I also see this issue. We (currently) don't allow external access to our source code so hooking into the AWS CI/CD workflow isn't really a fit for us at the moment.
> 
> > Is it possible to have one environment (dev) that only uses s3 and http and another enviornment (prod) that uses cloudfront and https? I have created different environments but if I do "amplify hosting configure" in one environment the changes effect the other environment as well

You could use our headless scripts in your own CI/CD system: https://github.com/aws-amplify/amplify-cli/tree/master/packages/amplify-cli/sample-headless-scripts

**Comment by jedlau** [comment]
> > Is it possible to have one environment (dev) that only uses s3 and http and another enviornment (prod) that uses cloudfront and https? I have created different environments but if I do "amplify hosting configure" in one environment the changes effect the other environment as well
> 
> @AshkanHovold Starting from your `dev env:
> 
> 1. `amplify env add` create`prod` which is a clone of `dev` at this point.
> 2. `amplify hosting remove`
> 3. `amplify hosting add` and add CloudFront this time round.
> 
> Let me know if that works.

@swaminator , This didn't work for me. I removed hosting from `prod`, added hosting back with CloudFront, but when I switched back to `dev`, that environment also wanted to update to CloudFront hosting.

**Comment by AshkanHovold** [comment]
Yes I have managed to have s3 on my "dev" environment and cloudfront on my "production" environment. If I publish production with cloudfront and then switch to my dev environment I have to do 
"amplify configure hosting" and remove cloudfront from there. If I publish it there then switch to production again, I need to re-add cloudfront there by doing "amplify configure hosting" again. It is very easy to miss this.
Imo this feels like a bugg but from amplify guides I have found it looks like this is by design? I would think having different environments with different setup is desired?
